### PR TITLE
Compute calibration coefficients using Vanadium

### PR DESCRIPTION
**Pull request by mganeva**
Algorithm to compute detector sensitivity calibration coefficients using Vanadium workspace. Developed for TOFTOF@MLZ, but can be used for other instruments if needed. Can be tested on sample workspaces (created by CreateSampleWorkspace, see unittests).

Please, review the code, tests and documentation. Let me know if I should improve them.

label: Direct inelastic

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+        # calculate array of Debye-Waller factors
+        dwf = self.calculate_dwf()
+
+        # for each detector: fit gaussian to get peak_centre and fwhm
+        # sum data in the range [peak_centre - 3*fwhm, peak_centre + 3*fwhm]
+        dataX = self.vanaws.readX(0)
+        coefY = np.zeros(nhist)
+        coefE = np.zeros(nhist)
+        for idx in range(nhist):
```
Comment (AndreiSavici) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: You should check if the spectra is masked

Comment (mganeva) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Good point, thanks.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                idxmin = (np.fabs(dataX-peak_centre+3.*fwhm)).argmin()
+                idxmax = (np.fabs(dataX-peak_centre-3.*fwhm)).argmin()
+                coefY[idx] = dwf[idx]*sum(dataY[idxmin:idxmax+1])
+                coefE[idx] = dwf[idx]*sum(dataE[idxmin:idxmax+1])
+
+        # create X array, X data are the same for all detectors, so
+        coefX = np.zeros(nhist)
+        coefX.fill(dataX[0])
+
+        api.CreateWorkspace(OutputWorkspace=outws_name, DataX=coefX, DataY=coefY, DataE=coefE, NSpec=nhist, UnitX="TOF")
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Would be best to do this through self.createChildAlgorithm. That way you'll have no need to go to the AnalysisDataService during the algorithm processing.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        outws = api.AnalysisDataService.retrieve(outws_name)
+        api.CopyLogs(self.vanaws, outws)
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Same as above regarding Algorithm generation.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+
+class ComputeCalibrationCoefVan(PythonAlgorithm):
+    """ Calculate coefficients to normalize by Vanadium and correct Debye Waller factor
+    """
+    def __init__(self):
+        """
+        Init
+        """
+        PythonAlgorithm.__init__(self)
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Not strictly necessary, but I can see why you do it.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        """
+        return "ComputeCalibrationCoefVan"
+
+    def summary(self):
+        return "Calculate coefficients for detector efficiency correction using the Vanadium data."
+
+    def PyInit(self):
+        """ Declare properties
+        """
+        self.declareProperty(WorkspaceProperty("VanadiumWorkspace", "", direction=Direction.Input,
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: You may want to consider being more restrictive on the WorkspaceProperty. I would imagine you really wanted to create a `MatrixWorkspaceProperty` not a `WorkspaceProperty`. Your algorithm shouldn't be able to accept an `ITableWorkspace` as input for example.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                                               validator=InstrumentValidator()),
+                             "Input Vanadium workspace")
+        self.declareProperty(WorkspaceProperty("OutputWorkspace", "", direction=Direction.Output),
+                             "Name the workspace that will contain the calibration coefficients")
+        return
+
+    def validateInputs(self):
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Great to see you've been using validateInputs.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+        if not run.hasProperty('wavelength'):
+            issues['VanadiumWorkspace'] = "Input workspace must have wavelength sample log."
+        else:
+            try:
+                float(run.getProperty('wavelength').value)
+            except ValueError:
+                issues['VanadiumWorkspace'] = "Invalid value for wavelength sample log. Wavelength must be a number."
+
+        outws = self.getPropertyValue("OutputWorkspace")
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Not necessary. The [PropertyMode](http://docs.mantidproject.org/nightly/api/python/mantid/api/PropertyMode.html) on the WorkspaceProperties defaults to `Mandatory` so you wouldn't get to this stage unless you had set the `OutputWorkspace`

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                                               validator=InstrumentValidator()),
+                             "Input Vanadium workspace")
+        self.declareProperty(WorkspaceProperty("OutputWorkspace", "", direction=Direction.Output),
+                             "Name the workspace that will contain the calibration coefficients")
+        return
+
+    def validateInputs(self):
+        issues = dict()
+        inws = self.getProperty("VanadiumWorkspace").value
+        run = inws.getRun()
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: `inws.getRun()` This will fail if you're `VanadiumWorkspace` input is not a subtype of `ExperimentInfo`. This is why you should fix your Input and Output workspace property types to be `MatrixWorkspaces`

File: docs/source/algorithms/ComputeCalibrationCoefVan-v1.rst [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+   where :math:`J(y) = 0.5` if :math:`y < 10^{-3}`, otherwise
+
+   :math:`J(y) = \int_0^1 x\cdot\mathrm{coth}\left(\frac{x}{2y}\right)\,\mathrm{d}x`
+
+   where :math:`y=T/T_m` is the ratio of the temperature during the experiment :math:`T` to the Debye temperature :math:`T_m = 389K`, :math:`m_V` is the Vanadium atomic mass (in kg) and :math:`\theta_i` is the polar angle of the i-th detector retrieved using the *detectorSignedTwoTheta* function.
+
+.. warning::
+
+    If sample log *temperature* is not present in the given Vanadium workspace or temperature is set to an invalid value, T=293K will be taken for the Debye-Waller factor calculation. Algorithm will produce warning in this case.
```
Comment (OwenArnold) [inline docs/source/algorithms/ComputeCalibrationCoefVan-v1.rst]: Good docs!

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
+from mantid.api import PythonAlgorithm, AlgorithmFactory, WorkspaceProperty, Progress, InstrumentValidator
+from mantid.kernel import Direction
+import mantid.simpleapi as api
+import numpy as np
+from scipy import integrate
+import scipy as sp
+import mlzutils
```
Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Not so great to see that we have to import *utilities* from a specific facility to get what looks like a general algorithm working, especially since `mlzutils.do_fit_gaussian` doesn't seem to really be facility specific. Or is it?

**Comment by OwenArnold** [comment]
@mganeva I've had a quick look at your PR. It mostly looks to be in order. I've made a few comments, but there's noting too major there. As these issues mostly seem minor, let me know if you need me to merge this badly.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py (hunk 10 above)
Comment (mganeva) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: No, it is not facility specific. Other algorithms in the mlzutils either. I've called it mlzutils because it is used now mainly for DNS and TOFTOF@MLZ, but I can rename it.  How to name it? May be utils?

Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: To avoid having Algorithms too intertwined with Mantid's persistent storage mechanism (The AnalysisDataService, better known as workspace list: Algorithms should never directly access any workspace information through the AnalysisDataService, and should never modify AnalysisDataService directly. This kind of direct access bypasses workspace read-write locking, can lead to race conditions between algorithms, and will make it very hard in the future to have session-based AnalysisDataServices. Seeing some of the use cases for mlzutils such as [DNSDetEffCorrVana](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/plugins/algorithms/DNSDetEffCorrVana.py#L96:L129) I suggest all operations in the mlzutils that deal with clearing and querying workspace lists are refactored away and removed.

Some of the functions you have in the mlzutils look genuinely useful elsewhere `compare_properties` and `compare_mandatory` are related and to do with property comparison. So either, create a module called PropertyComparitor which your algorithm can use, or create a base class derived from `PythonAlgorithm`, that future and existing `Algorithms` can derive from. `do_gaussian_fit` only seems to be used in this *ComputeCalibrationCoefVan.py* at the moment, so I suggest you add it as a member function to that Algorithm until you have a need to use it else where and then follow the rules to make it either a module or base Algorithm.

In summary we are now trying very hard on the Mantid project to name anything as a *utility* *helper* or *manager*, as the clients of these tools have no idea what's in them based on the name and they end up becoming dumping grounds for unstructured code.

This is a rather long response. I'm happy to merge your code as is, if you like, but if you can clean up the mlzutilities in the near future as I have described above, I would be very grateful.

Comment (mganeva) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: I will refactor mlzutils. May be it would be better to submit a separate refactoring pull request, since several algorithms will get involved. However, do_gaussian_fit is also used in the other TOFTOF  algorithm which I am preparing for submission now. I've placed it in the mlzutils to avoid the code duplication. What would be better to do with do_gaussian_fit? May be also create a generic module?

Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: Yes a separate pull request with the refactoring changes will be fine. Yes a separate module containing the do_guassian_fit would probably be best.

File: Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py (hunk 2 above)
Comment (mganeva) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: I have a problem to execute "CreateWorkspace" as a child algorithm. Code in PyExec:
        create = self.createChildAlgorithm("CreateWorkspace")
        create.setPropertyValue('OutputWorkspace', outws_name)
        create.setProperty('DataX', coefX)
        create.setProperty('DataY', coefY)
        create.setProperty('DataE', coefE)
        create.setProperty('NSpec', nhist)
        create.setProperty('UnitX', 'TOF')
        status = create.execute()
        outws = create.getProperty('OutputWorkspace').value
        self.log().warning("Outws: " + outws.getName())
        self.log().warning("Status: " + str(status))
Gives the following output:
ComputeCalibrationCoefVan-[Warning] Outws: 
ComputeCalibrationCoefVan-[Warning] Status: True
I don't understand why the name of the created workspace is an empty string. What I am doing wrong?

Comment (OwenArnold) [inline Framework/PythonInterface/plugins/algorithms/ComputeCalibrationCoefVan.py]: You're not doing anything wrong. Your workspace does not have a name because only things that go in the AnalysisDataService dictionary are given names, and you're no longer putting it in the AnalysisDataService as a side-effect. The workspace does exist, but no workspace will have a name yet. When your full Algorithm (ComputeCalibrationCoefVan) completes, the OutputWorkspace will appear in the AnalysisDataService with the name that you set.

**Comment by mganeva** [comment]
@OwenArnold I have not changed the mlzutils module. I will submit the refactoring as a separate pull request. Thank you for reviewing.
